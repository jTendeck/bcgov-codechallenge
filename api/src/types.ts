//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

export interface Options {
    baseURL: string;
}

export interface QueriedPoint {
    longitude: number;
    latitude: number;
}

export interface ChsaResponseSet {
    CMNTY_HLTH_SERV_AREA_CODE: string;
    CMNTY_HLTH_SERV_AREA_NAME: string;
}

export interface ChsaFeature {
    type: string;
    id: string;
    geometry: any;
    properties: {
        HLTH_CHSA_SYSID: number;
        CMNTY_HLTH_SERV_AREA_CODE: string;
        CMNTY_HLTH_SERV_AREA_NAME: string;
        OBJECTID: number;
    }
}

export interface ChsaPayload {
    type: string;
    features: ChsaFeature[];
    totalFeatures: number;
    numberMatched: number;
    numberReturned: number;
    timeStamp: Date;
    crs: any;
}
