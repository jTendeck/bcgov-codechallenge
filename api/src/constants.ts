//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

export const ENVIRONMENTS = {
  DEVELOPMENT: 'development',
  PRODUCTION: 'production',
};

export const GEO_API_CALL = {
  SERVICE: 'WFS',
  VERSION: '1.0.0',
  REQUEST: 'GetFeature',
  TYPE_NAME: 'pub:WHSE_ADMIN_BOUNDARIES.BCHA_CMNTY_HEALTH_SERV_AREA_SP',
  SRS_NAME: 'EPSG:4326',
  PROPERTY_NAME: 'CMNTY_HLTH_SERV_AREA_CODE,CMNTY_HLTH_SERV_AREA_NAME',
  OUTPUT_FORMAT: 'application/json',
};
