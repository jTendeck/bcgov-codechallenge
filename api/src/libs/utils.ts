//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';

// extra check_empty function is needed
// because lodash isEmpty seems to produce confusing results
// e.g. lodash.isEmpty(true) OR lodash.isEmpty(1) yields true
export const isEmptyValue = (value: any): boolean => {
  return value === undefined || value === null || value === ''
    || (typeof value === 'object' && Object.keys(value).length === 0)
};

export const isValidLongitude = (longitude: number): boolean => {
  if (typeof longitude !== 'number') {
    throw new Error('invalid argument to isValidLongitude function');
  }

  return isFinite(longitude) && Math.abs(longitude) <= 180;
};

export const isValidLatitude = (latitude: number): boolean => {
  if (typeof latitude !== 'number') {
    throw new Error('invalid argument to isValidLatitude function');
  }

  return isFinite(latitude) && Math.abs(latitude) <= 90;
};

export const convertNumToStrFormatForGeoApi = (value: number): string => {
  if (typeof value !== 'number') {
    throw new Error('invalid argument to convertNumToStrFormatForGeoApi function');
  }

  if (value >= 0) {
    return `+${value.toString()}`;
  } else {
    return value.toString();
  }
};

export const writeToDbSuccessCallback = (log: any): void => {
  try {
    logger.info(`successfully wrote log (${JSON.stringify(log)})`);
  } catch (err) {
    logger.error(`unable to process writeToDbSuccessCallback routine`);
  }
};
