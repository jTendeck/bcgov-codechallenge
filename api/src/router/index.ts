//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import cors from 'cors';
import config from '../config';
import chsa from './routes/chsa';
import ehlo from './routes/ehlo';

const corsOptions = {
  origin: config.get('environment') === 'development' ? '*' : config.get('apiUrl'),
  credentials: true,
  optionsSuccessStatus: 200, // some legacy browsers (IE11, various SmartTVs) choke on 204
};

export const router = app => {
  app.use(cors(corsOptions));
  app.use('/api/v1/ehlo', ehlo); // probes
  app.use('/api/v1/chsa', chsa); // chsa resources
};
