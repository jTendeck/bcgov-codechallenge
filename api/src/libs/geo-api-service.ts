//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import axios, { AxiosInstance } from 'axios';
import { GEO_API_CALL } from '../constants';
import { ChsaFeature, ChsaPayload, ChsaResponseSet, Options, QueriedPoint } from '../types';
import { convertNumToStrFormatForGeoApi } from './utils';

export default class GeoApiService {
  private axi: AxiosInstance;

  constructor(options: Options) {
    this.axi = axios.create({
      baseURL: options.baseURL,
    });
  }
  public async fetchChsaResponseSet(queriedPoint: QueriedPoint): Promise<ChsaResponseSet> {
    const { longitude, latitude } = queriedPoint;
    const longStr = convertNumToStrFormatForGeoApi(longitude);
    const latStr = convertNumToStrFormatForGeoApi(latitude);

    // TOODO: investigate the reason when cql_filter field is included in params
    // axios call somehow cause geo api throw an exception re parsing cql_filter
    const params = {
      service: GEO_API_CALL.SERVICE,
      version: GEO_API_CALL.VERSION,
      request: GEO_API_CALL.REQUEST,
      typeName: GEO_API_CALL.TYPE_NAME,
      srsname: GEO_API_CALL.SRS_NAME,
      propertyName: GEO_API_CALL.PROPERTY_NAME,
      outputFormat: GEO_API_CALL.OUTPUT_FORMAT,
    };

    try {
      const response = await this.axi.get(`ows?cql_filter=INTERSECTS(SHAPE,SRID=4326;POINT(${longStr}${latStr}))`, {
        headers: {
          'content-type': 'application/json',
        },
        params,
      });

      // TODO: implement a solution for 'content-type' / 'Content-Type' issue

      // when geo api returns 200 but with a static page that shows certain exceptions
      const contentType = response.headers['content-type'];
      if (!(contentType && contentType.includes('application/json'))) {
        throw errorWithCode('unable to query chsa response set', 500);
      }

      const payload: ChsaPayload = response.data;
      logger.info(`successfully received chsa payload (${JSON.stringify(payload)})`);
      const features: ChsaFeature[] = payload.features;

      // when geo api receives syntactically valid point but is outside of BC
      if (features.length === 0) {
        throw errorWithCode(`invalid longtitude and latitude for a BC location,\
        you can use google map to search for the correct and try again `, 422);
      }

      // when geo api receives valid point
      const { properties: { CMNTY_HLTH_SERV_AREA_CODE, CMNTY_HLTH_SERV_AREA_NAME } } = features[0];
      return { CMNTY_HLTH_SERV_AREA_CODE, CMNTY_HLTH_SERV_AREA_NAME };
    } catch (err) {
      return errorWithCode(err.message, err.code);
    }
  }
}
