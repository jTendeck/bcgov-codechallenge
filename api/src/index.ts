//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';
import bodyParser from 'body-parser';
import flash from 'connect-flash';
import cookieParser from 'cookie-parser';
import express from 'express';
import fs from 'fs';
import path from 'path';
import { router } from './router';

const app = express();
const options = {
  inflate: true,
  limit: '204800kb', // 200Mb
  type: 'image/*',
};
const docpath = path.join(__dirname, '../', 'public/doc/api');
const pubpath = path.join(__dirname, '../', 'public');

fs.access(docpath, fs.constants.R_OK, err => {
  if (err) {
    logger.warn('API documentation does not exist');
    return;
  }

  app.use('/doc', express.static(docpath));
});

fs.access(pubpath, fs.constants.R_OK, err => {
  if (err) {
    logger.warn('static assets location does not exist');
    return;
  }

  app.use('/', express.static(pubpath));
});

app.use(cookieParser());
app.use(
  bodyParser.urlencoded({
    extended: true,
  })
);
app.use(bodyParser.json());
app.use(bodyParser.raw(options));
app.use(flash());

// server api routes
router(app);

// last rescue for error handling
// this needs to be last in or it will not get called
// eslint-disable-next-line no-unused-vars
app.use((err, req, res, next) => {
  logger.error(err.message);
  const code = err.code ? err.code : 500;
  const message = err.message ? err.message : 'internal server error';

  res.status(code).json({ error: message, success: false });
});

export default app;
