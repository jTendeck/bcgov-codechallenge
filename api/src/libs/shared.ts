//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { Pool } from 'pg';
import config from '../config';
import { Options } from '../types';
import { default as GeoApiService } from './geo-api-service';

interface Shared {
  pgPool: Pool;
  geo: GeoApiService;
}

const pgPoolKey = Symbol.for('chsa.pgpool');
const geoKey = Symbol.for('chsa.geo');

const gs = Object.getOwnPropertySymbols(global);

const main = async () => {
  if (gs.indexOf(geoKey) <= -1) {
    const opts: Options = {
      baseURL: config.get('geo:baseURL'),
    };

    const geo = new GeoApiService(opts);
    global[geoKey] = geo;
  }

  if (gs.indexOf(pgPoolKey) <= -1) {
    const params = {
      host: config.get('db:host'),
      port: config.get('db:port'),
      database: config.get('db:database'),
      user: config.get('db:user'),
      password: config.get('db:password'),
      max: config.get('db:maxConnections'),
      idleTimeoutMillis: config.get('db:idleTimeout'),
      connectionTimeoutMillis: config.get('db:connectionTimeout'),
    }

    global[pgPoolKey] = new Pool(params);
  }
}

main();

const shared = {};

Object.defineProperty(shared, 'geo', {
  get: () => global[geoKey],
});

Object.defineProperty(shared, 'pgPool', {
  get: () => global[pgPoolKey],
});

Object.freeze(shared);

export default shared as Shared;
