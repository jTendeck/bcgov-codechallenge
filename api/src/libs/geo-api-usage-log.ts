//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';
import DataManager from '../db';
import { GeoApiUsageLog } from '../db/model/geo-api-usage-log';
import { ChsaResponseSet, QueriedPoint } from '../types';
import shared from './shared';

const dm = new DataManager(shared.pgPool);
const { GeoApiUsageLogModel } = dm;

export const writeGeoApiUsageLog = async (queriedPoint: QueriedPoint, chsaResponseSet: ChsaResponseSet | any): Promise<GeoApiUsageLog> => {
    try {
        const { CMNTY_HLTH_SERV_AREA_CODE, CMNTY_HLTH_SERV_AREA_NAME } = chsaResponseSet;
        const { latitude, longitude } = queriedPoint;

        return await GeoApiUsageLogModel.create({
            queriedLongitude: longitude,
            queriedLatitude: latitude,
            resultChsaCode: CMNTY_HLTH_SERV_AREA_CODE,
            resultChsaName: CMNTY_HLTH_SERV_AREA_NAME,
        });
    } catch (err) {
        const message = `unable to write geo api usage log for querying ${JSON.stringify(queriedPoint)}`;
        logger.error(`${message}, err = ${err.message}`);

        throw err;
    }
};
