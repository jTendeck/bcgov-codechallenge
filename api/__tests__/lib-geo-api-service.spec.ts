//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import mockAxios from 'axios';
import fs from 'fs';
import path from 'path';
import { default as GeoApiService } from '../src/libs/geo-api-service';
import { Options } from '../src/types';

const p0 = path.join(__dirname, 'fixtures/get-geo-api-response-invalid-bc-location.json');
const invalidBCLocationResponse = JSON.parse(fs.readFileSync(p0, 'utf8'));

const p1 = path.join(__dirname, 'fixtures/get-geo-api-response-valid-bc-location.json');
const validBCLocationResponse = JSON.parse(fs.readFileSync(p1, 'utf8'));

describe('Services', () => {
  const options: Options = {
    baseURL: 'something',
  };
  const geo = new GeoApiService(options);
  const headers = {
    'content-type': 'application/json'
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  // TODO - create a test that uses const invalidBCLocationResponse
  // it('queryChsaResponseSet() works correctly for invalid BC location', async () => {
  //
  // });

  it('queryChsaResponseSet() works correctly for valid BC location', async () => {
    const queriedPoint = {
      longitude: -123.711,
      latitude: 48.8277,
    };
    // @ts-ignore
    mockAxios.fn.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: validBCLocationResponse,
        headers,
      })
    );

    const result = await geo.fetchChsaResponseSet(queriedPoint);

    expect(result).toMatchSnapshot();
    // @ts-ignore
    expect(mockAxios.fn.get).toHaveBeenCalledTimes(1);
  });

  it('queryChsaResponseSet() works correctly when the external api responds non success status code', async () => {
    const queriedPoint = {
      longitude: -123.711,
      latitude: 48.8277,
    };
    // @ts-ignore
    mockAxios.fn.post.mockImplementationOnce(() =>
      Promise.reject()
    );

    const result = await geo.fetchChsaResponseSet(queriedPoint);
    expect(result).toMatchSnapshot();
  });

  it('queryChsaResponseSet() works correctly when the external api responds non json', async () => {
    const queriedPoint = {
      longitude: -123.711,
      latitude: 48.8277,
    };

    // @ts-ignore
    mockAxios.fn.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: validBCLocationResponse,
        headers: {
          'content-type': 'text/xml;charset=UTF-8',
        }
      })
    );

    const result = await geo.fetchChsaResponseSet(queriedPoint);
    expect(result).toMatchSnapshot();
  });
});
