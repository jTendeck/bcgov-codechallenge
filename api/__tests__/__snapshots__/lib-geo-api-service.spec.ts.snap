// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Services queryChsaResponseSet() works correctly for invalid BC location 1`] = `[Error: invalid longtitude and latitude for a BC location,        you can use google map to search for the correct and try again ]`;

exports[`Services queryChsaResponseSet() works correctly for valid BC location 1`] = `
Object {
  "CMNTY_HLTH_SERV_AREA_CODE": "4212",
  "CMNTY_HLTH_SERV_AREA_NAME": "Central Cowichan",
}
`;

exports[`Services queryChsaResponseSet() works correctly when the external api responds non json 1`] = `[Error: unable to query chsa response set]`;

exports[`Services queryChsaResponseSet() works correctly when the external api responds non success status code 1`] = `[Error: Cannot read property 'content-type' of undefined]`;
